lib/logstash/outputs/tcp.rb

```diff
@@ -104,50 +109,73 @@ def setup_ssl
       @ssl_context.cert_store = @cert_store
       @ssl_context.verify_mode = OpenSSL::SSL::VERIFY_PEER|OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT
     end
-  end # def setup_ssl
 
-  public
+    @ssl_context.min_version = :TLS1_1 # not strictly required - JVM should have disabled TLSv1
+    if ssl_supported_protocols.any?
+      disabled_protocols = ['TLSv1.1', 'TLSv1.2', 'TLSv1.3'] - ssl_supported_protocols
+      unless OpenSSL::SSL.const_defined? :OP_NO_TLSv1_3 # work-around JRuby-OpenSSL bug - missing constant
+        @ssl_context.max_version = :TLS1_2 if disabled_protocols.delete('TLSv1.3')
+      end
+      # mapping 'TLSv1.2' -> OpenSSL::SSL::OP_NO_TLSv1_2
+      disabled_protocols.map! { |v| OpenSSL::SSL.const_get "OP_NO_#{v.sub('.', '_')}" }
+      @ssl_context.options = disabled_protocols.inject(@ssl_context.options) { |options, op_no| options | op_no }
+    end
+    @ssl_context
+  end
+  private :setup_ssl
+
+  # @note to be able to hook up into #ssl_context from tests
+  def new_ssl_context
+    OpenSSL::SSL::SSLContext.new
+  end
+  private :new_ssl_context
+
+  # @overload Base#register
   def register
     require "socket"
     require "stud/try"
-    if @ssl_enable
-      setup_ssl
-    end # @ssl_enable
+    @closed = Concurrent::AtomicBoolean.new(false)
+    setup_ssl if @ssl_enable
 
     if server?
       @logger.info("Starting tcp output listener", :address => "#{@host}:#{@port}")
       begin
         @server_socket = TCPServer.new(@host, @port)
       rescue Errno::EADDRINUSE
-        @logger.error("Could not start TCP server: Address in use",
-                      :host => @host, :port => @port)
+        @logger.error("Could not start tcp server: Address in use", host: @host, port: @port)
         raise
       end
       if @ssl_enable
         @server_socket = OpenSSL::SSL::SSLServer.new(@server_socket, @ssl_context)
       end # @ssl_enable
-      @client_threads = []
+      @client_threads = Concurrent::Array.new
 
       @accept_thread = Thread.new(@server_socket) do |server_socket|
+        LogStash::Util.set_thread_name("[#{id}]|output|tcp|server_accept")
         loop do
-          Thread.start(server_socket.accept) do |client_socket|
+          break if @closed.value
+          client_socket = server_socket.accept_nonblock exception: false
+          if client_socket == :wait_readable
+            IO.select [ server_socket ]
+            next
+          end
```

Comment:
Here (may)be dragons. While I think this is correct I'd like to review it a little more carefully tomorrow as we've had bad experiences before with (j)ruby's nonblock+select before

Reply:
the reasoning here is that we're not in the blocking `server_socket.accept` - cause that will raise when close-ing the plugin. believe the use of `logstash-code >= 8.1` guarantees an up-to-date JRuby 9.2.x as the minimum but sure if you feel there are issues I am happy to switch back to a blocking version.